pie chart labels

ppisljar opened the pull request:
Release Notes: Users can now add labels to pie charts.

![image](https://user-images.githubusercontent.com/1833023/33722463-0b0877aa-db38-11e7-9c03-684fb84b23f2.png)


--------------------------------------------

Resolves #1702 
partly addresses: #3686 

summary: pie charts can now have labels on them to increase readability

ppisljar commented:
jenkins, test this

thomasneirynck commented on src/ui/public/vislib/visualizations/pie_chart.js:
```diff
@@ -165,6 +188,165 @@ export function VislibVisualizationsPieChartProvider(Private) {
+            };
+
+            const conflicts = [];
+            let conflictIterations = 0;
```
this algo is a little dense for me. might help to go over with me in person.

ppisljar commented:
thanks for looking into this @thomas ... answers to some of your comments
- fixed the bug with single label

- missing labels on the right side were not related to conflict resolution, but were simply pushed out too far to the left (and then hidden as they would go out of the vis bounds) ... i updated this to show labels closer to the pie itself, however you can still get into that scenario (on dashboard for example you should not make your container square .... but 'widescreen' so it can fit the labels as well.

- colors of lines and texts: we loose legibility quite fast (think of light yellow slices)

- truncation .... i would leave it as it is in this PR for consistency reasons (its like this everywhere else we have it). but i agree we should update this (everywhere).

ppisljar commented:
flipping of elbows: we don't flip them ... thats the thing ... elbow is always drawn thru 3 points:
- centroid of the slice
- middle point where x = centroid of outer arc and y = label y
- label position

doesn't work well in all scenarios ... but not sure its worth complicating, it will mostly be used for simple charts probably.

thomasneirynck commented:
I didn't explain well enough about the "elbows". I mean, the elbow flips when you move labels that connect one quadrant of the pie to another quadrant. that's how you get that overlap.

e.g.: when there's a pie slice in the bottom-left quadrant, but position the label in the top-left quadrant, the "elbow" flips, and the connector-line crosses the chart.

agreed on preserving the black color, doing otherwise would make it worse

ppisljar commented:
@thomas i simplified conflict resolution a bit ... now elbows wont 'break' ... simply if label cant be positioned without conflict its skipped .... 

this will render less labels, and won't prioritize labels with bigger slice size over the labels with smaller slices ... let me know what you think

manninge commented:
So I see this has been solved, but I do not see the 'how to'. I am new to Kibana, so a little green. Picking up a project dumped by someone else and I need to add labels to the pie slices. Thanks.

thomasneirynck commented:
Created https://github.com/elastic/kibana/pull/14926 to solve merge issues.

timroes commented:
Close this in favor of #14926

ppisljar commented:
@timroes all your comments were addressed (from #14926)
@alexfrancoeur almoust all your comments were addressed ... for now i didn't add another checkbox to actually show the value in labels (instead of percentage) ... i guess we can do that as a follow up, when somebody opens an issue for it.

ppisljar commented:
CI failed on `fail: "management import objects should allow for cancelling overrides"` which i think is unrelated ... restarting

jenkins, test this

ppisljar commented:
CI failed on `fail: "visualize app tile map visualize app tile map chart should save with zoom level and load, take screenshot"` which is also unrelated ....

you just have to love our CI :)

jenkins, test this

thomasneirynck commented:
when merging, could you also add a small `summary` paragraph? thx!

ppisljar commented:
backport 6.x: 0b5052a

alexfrancoeur commented:
yes @ppisljar!!! 🍻

gui-f commented:
Great news! 

However, I believed that on this version the # of counts of documents within legends of certain chart types (like bar and pie) would be available again.

![image](https://user-images.githubusercontent.com/35538747/35059380-176c9a96-fba2-11e7-93f8-04d5ee0fb4e2.png)

Is there any way to include these informations on charts with terms aggregation?

ppisljar commented:
you are right, currently you can only show the percentage value, but not the actual document count.
could you open a separate issue for that ? thanks!

cawoodm commented:
I'm still seeing no values on my Pie chart if I disable Labels (which I don't want) but enable Values (which I do want):

![image](https://user-images.githubusercontent.com/498834/83226618-a8522e80-a182-11ea-9031-eb960c4e8cc7.png)

![image](https://user-images.githubusercontent.com/498834/83226628-ae480f80-a182-11ea-9611-cc5ec8909a23.png)
